src/pepr/operator/controllers/istio/egress-ambient.ts

```diff
@@ -34,72 +81,203 @@ export async function applyAmbientEgressResources(packageList: Set<string>, gene
+      },
+      "Resolved ambient egress identities for host",
     );
+    if (saPrincipals.length === 0 && namespaces.length === 0) {
+      log.warn(
+        { host },
+        "Skipping ambient egress resources for host due to empty source identities (owners/participants not yet resolved)",
+      );
+      continue;
+    }
+
+    // Apply shared Ambient ServiceEntry only when we have at least one source identity
+    const se = generateSharedAmbientServiceEntry(host, resource, generation);
+    log.debug(se, `Applying Ambient Shared ServiceEntry ${se.metadata?.name}`);
+    await K8s(IstioServiceEntry).Apply(se, { force: true });
+
+    const ap = generateCentralAmbientEgressAuthorizationPolicy(host, generation, {
+      saPrincipals,
+      namespaces,
+    });
+    log.debug(ap, `Applying Ambient Central AuthorizationPolicy ${ap.metadata?.name}`);
+    await K8s(IstioAuthorizationPolicy).Apply(ap, { force: true });
+  }
+
+  // Cleanup: purge legacy local egress resources created prior to ambient changes
+  // Criteria per package (in its namespace):
+  // - ServiceEntry with labels: uds/package=<pkgName> AND istio.io/use-waypoint=egress-waypoint
+  // - AuthorizationPolicy with labels: uds/package=<pkgName> AND uds/for=egress
+  // Only for packages that are Ambient and contributed to this central reconcile
+  try {
+    const contribIds = new Set(Array.from(packageList)); // ids are `${name}-${namespace}`
+    for (const pkg of pkgItems) {
+      const ns = pkg.metadata?.namespace;
+      const name = pkg.metadata?.name;
+      const mode = pkg.spec?.network?.serviceMesh?.mode || Mode.Sidecar;
+      if (!ns || !name || mode !== Mode.Ambient) continue;
+      const id = `${name}-${ns}`;
+      if (!contribIds.has(id)) continue;
 
-    log.debug(authPolicy, `Applying Authorization Policy ${authPolicy.metadata?.name}`);
+      await purgeOrphans(String(generation), ns, name, IstioServiceEntry, log, {
+        "istio.io/use-waypoint": "egress-waypoint",
+      });
+      await purgeOrphans(String(generation), ns, name, IstioAuthorizationPolicy, log, {
+        "uds/for": "egress",
+      });
+    }
+  } catch (e) {
+    log.warn(
+      { err: e },
+      "Failed purging legacy local egress resources for ambient packages (non-fatal)",
+    );
+  }
+}
 
-    // Apply the AuthorizationPolicy and force overwrite any existing resource
-    await K8s(IstioAuthorizationPolicy).Apply(authPolicy, { force: true });
+// Purge any orphaned ambient egress resources
+export async function purgeAmbientEgressResources(generation: string) {
+  try {
+    await purgeOrphans(generation, ambientEgressNamespace, sharedEgressPkgId, K8sGateway, log);
+    await purgeOrphans(
+      generation,
+      ambientEgressNamespace,
+      sharedEgressPkgId,
+      IstioServiceEntry,
+      log,
+    );
+    await purgeOrphans(
+      generation,
+      ambientEgressNamespace,
+      sharedEgressPkgId,
+      IstioAuthorizationPolicy,
+      log,
+    );
```

Comment:
I'm having trouble tracing down the "why" but this purge seems to have a race condition or other bug where it can result in most resources getting purged. I've been testing with a [large set of packages](https://gist.github.com/mjnagel/3cdf18684f32726fe48a680b643b3664) and when applying them all together I end up with no egress service entries or authpols in my cluster. Restarting the watcher typically works to create them, but there's also some strange behavior on delete at times where deleting all the packages together sometimes creates/leaves them.